src/lib/parallel_scan/parallel_scan.ml

```diff
@@ -1,1328 +1,1367 @@
   in
-  Option.is_some (State.base_none_pos t)
-  && free_space ~state:t > 0
-  && has_valid_merge_jobs && non_empty_tree && validate_leaves
-  && validate_levels
-
-let fill_in_completed_jobs :
-       state:('a, 'd) State.t
-    -> completed_jobs:'a State.Completed_job.t list
-    -> ('a * 'd list) option Or_error.t =
- fun ~state ~completed_jobs ->
-  let open Or_error.Let_syntax in
-  let old_jobs = Ring_buffer.copy state.jobs in
-  let last_acc = state.acc in
-  let%map () = State.consume state completed_jobs old_jobs in
-  state.jobs.position <- 0 ;
-  if not (fst last_acc = fst state.acc) then snd state.acc else None
+  let%map _ =
+    State_monad.put
+      {state with trees= append updated_trees (Non_empty_list.tail state.trees)}
+  in
+  ()
+
+let incr_sequence_no = function
+  | state ->
+      let open State_monad in
+      (*let open State_monad.Let_syntax in*)
+      put {state with curr_job_seq_no= state.curr_job_seq_no + 1}
+
+let update_helper :
+    data:'d list -> completed_jobs:'a list -> ('b, 'a, 'd) State_monad.t =
+ fun ~data ~completed_jobs ->
+  let open State_monad in
+  let open State_monad.Let_syntax in
+  let%bind t = get in
+  let%bind () =
+    error_if
+      (List.length data > t.max_base_jobs)
+      ~message:
+        (sprintf
+           !"Data count (%d) exceeded maximum (%d)"
+           (List.length data) t.max_base_jobs)
+  in
+  let delay = t.delay + 1 in
+  (*Increment the sequence number*)
+  let%bind () = incr_sequence_no t in
+  let latest_tree = Non_empty_list.head t.trees in
+  let available_space = Tree.required_job_count latest_tree in
+  (*Possible that new base jobs be added to a new tree within an update. This happens when the throughput is not always at max. Which also requires merge jobs to be done one two different set of trees*)
+  let data1, data2 = List.split_n data available_space in
+  let required_jobs_for_current_tree =
+    work (Non_empty_list.tail t.trees) ~max_base_jobs:t.max_base_jobs ~delay
+    |> List.length
+  in
+  let jobs1, jobs2 =
+    List.split_n completed_jobs required_jobs_for_current_tree
+  in
+  (*update fist set of jobs and data*)
+  let%bind result_opt = add_merge_jobs ~completed_jobs:jobs1 in
+  let%bind () = add_data ~data:data1 in
+  (*update second set of jobs and data. This will be empty if all the data fit in the current tree*)
+  let%bind _ = add_merge_jobs ~completed_jobs:jobs2 in
+  let%bind () = add_data ~data:data2 in
+  (*Check the tree-list length is under max*)
+  let%bind state = State_monad.get in
+  let%map () =
+    error_if
+      (Non_empty_list.length state.trees > max_trees state)
+      ~message:
+        (sprintf
+           !"Tree list length (%d) exceeded maximum (%d)"
+           (Non_empty_list.length state.trees)
+           (max_trees state))
+  in
+  result_opt
 
-let last_emitted_value (state : ('a, 'd) State.t) = snd state.acc
+let update :
+       data:'d list
+    -> completed_jobs:'a list
+    -> ('a, 'd) t
+    -> (('a * '
```

Comment:
a lot of tests are deleted here, can more of them be converted over or is there an impedance mismatch now?